v2/concordium/types.proto

```diff
@@ -3255,4 +3255,293 @@ message WinningBaker {
+        //
+        // The energy cost for this operation is 2000.
+        BlockHashInput load_block_state = 1;
+        // Run a query on the state.
+        DryRunStateQuery state_query = 2;
+        // Run a (non-transaction) operation to modify the state.
+        DryRunStateOperation state_operation = 3;
+    }
+}
+
+// Run a query as part of a dry run. Queries do not update the block state.
+message DryRunStateQuery {
+    oneof query {
+        // Look up information on a particular account.
+        //
+        // The energy cost for this query is 200.
+        AccountIdentifierInput get_account_info = 1;
+        // Look up information about a particular smart contract.
+        //
+        // The energy cost for this query is 200.
+        ContractAddress get_instance_info = 2;
+        // Invoke an entrypoint on a smart contract instance.
+        // No changes made to the state are retained at the completion of the operation.
+        //
+        // The energy cost for this query is 200 plus the energy used by the smart contract
+        // execution.
+        DryRunInvokeInstance invoke_instance = 3;
+    }
+}
+
+// Invoke an entrypoint on a smart contract instance.
+message DryRunInvokeInstance {
+    // Invoker of the contract. If this is not supplied then the contract will be
+    // invoked by an account with address 0, no credentials and sufficient amount
+    // of CCD to cover the transfer amount. If given, the relevant address (either
+    // account or contract) must exist in the blockstate.
+    optional Address invoker = 1;
+    // Address of the contract instance to invoke.
+    ContractAddress instance = 2;
+    // Amount to invoke the smart contract instance with.
+    Amount amount = 3;
+    // The entrypoint of the smart contract instance to invoke.
+    ReceiveName entrypoint = 4;
+    // The parameter bytes to include in the invocation of the entrypoint.
+    Parameter parameter = 5;
+    // The maximum energy to allow for the invocation. Note that the node imposes an energy
+    // quota that is enforced in addition to this limit.
+    Energy energy = 6;
+}
+
+// An operation that can update the state as part of a dry run.
+message DryRunStateOperation {
+    oneof operation {
+        // Sets the current block time to the given timestamp for the purposes of future
+        // transactions.
+        //
+        // The energy cost of this operation is 50.
+        Timestamp set_timestamp = 1;
+        // Add a specified amount of newly-minted CCDs to a specified account.
+        // The amount cannot cause the total circulating supply to overflow.
+        //
+        // The energy cost of this operation is 400.
+        DryRunMintToAccount mint_to_account = 2;
+        // Dry run a transaction, updating the state if it succeeds.
+        //
+        // The energy cost of this operation is 400 plus the energy used by executing the
+        // transaction.
+        DryRunTransaction run_transaction = 3;
```

Comment:
I can see you did not add account creation to the list which means we cannot replay transactions.

I think this feature has been going long enough so that I don't want to block merging this, but I am disappointed that was not done, since it's one of the two use-cases for this.

Reply:
See https://github.com/Concordium/concordium-node/pull/1051#issuecomment-1777196087

In the feature proposal, account creation and parameter update transactions were explicitly not supported, as well as executing block prologue/epilogue. It is perfectly possible to replay the other transactions in the block, and my understanding was that this would be sufficient.